src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -90,44 +90,89 @@ public Dictionary(int capacity, IEqualityComparer<TKey>? comparer)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.dictionary);
             }
 
+            Copy(dictionary, this);
+        }
+
+        public Dictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection) : this(collection, null) { }
+
+        public Dictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection, IEqualityComparer<TKey>? comparer) :
+            this((collection as ICollection<KeyValuePair<TKey, TValue>>)?.Count ?? 0, comparer)
+        {
+            if (collection == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.collection);
+            }
+
+            Copy(collection, this);
+        }
+
+        private static void Copy(IEnumerable<KeyValuePair<TKey, TValue>> collection, Dictionary<TKey, TValue> dest)
```

Comment:
Would this look better as an instance method?

Reply:
Yes; though are the `IEnumerable<KeyValuePair<TKey, TValue>>` and `IDictionary<TKey, TValue>` uncommon constructors.

Was wondering what can be done here for AoT; wasn't sure if having it as a static reduced the generic code gen when its not called?

Reply:
Static vs. instance private methods won't make a difference.

Reply:
If anything, it would be nice to just forward from the IDictionary constructor to IEnumerable constructor without introducing the extra method - but I am not sure what it would take.

Maybe it is best to leave IEnumerable constructor alone and not try to optimize. I think the chances for this optimization kicking in for that constructor are pretty low.